icefoganalytics/internal-data-portal
# User: All Datasets Page - User All Datasets Page Access And Actions

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/20

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7
- See [All Datasets page - /datasets](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit#heading=h.hyc5ymxijixo)

# Context

Out of Scope:
- add create, update, delete for data owner
- search

## All Datasets page - /datasets

### Back-end routes

* /datasets/search?term=... - magic search endpoint
    * Filter by tags/keywords
    * Filter by department
    * filter by ownership
* /datasets - generic paginated data sets list, not sortable at this time.


### Components

* `BaseLayout.vue` - see previous
* `DatasetSearch.vue` - Generates a search query against all datasets and redirects the user to a page with the appropriate fields filled in?
* `DatasetTable.vue` - list all datasets
    * Dataset - Dataset#name
    * Description - Dataset#description
    * Keywords - Tags through Taggings
    * Owner - Dataset#owner -> department
    * Access - magic aggregate field that describes the type of access the current user has available against the database, should be serialized in the back-end from the AccessGrants for the dataset.
    * (unlabeled) - aggregate of possible actions and states depending on available AccessGrants and current AccessRequests, and the type of access request (i.e. subscriptions). This is pretty complex logic and would benefit from testing.
        * `RequestAccessButton.vue`
            * Either opens a request access dialog or redirects to a request access form
        * `SubscribeToDatasetButton.vue`
            * Either opens a subscription request dialog or redirects to a subscription request form
        * Subscribed state
        * Approved state
        * Awaiting approval state

**Additional context**

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/2da7a524-3171-4fc3-a6b6-610619201428)

# Implementation

Break datasets serializers out into smaller classes, and implement a base serializer.
Add a bunch of tests.
Implement test error logging patch for Sequelize + Jest.

# Screenshots

http://localhost:8080/datasets
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/b5c2be19-c98a-4184-84f8-a3d10082f2d7)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/97ba4fc7-a5b7-4973-8318-e19ffaaf99ce)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. From the dashboard you see after login (http://localhost:8080/dashboard) apply some random access grants to your datasets, or create some if you don't have any.
5. Click the "View Datasets" button on the dashboard to be redirected to http://localhost:8080/datasets
6. Check that the "access" and "actions" columns look correct.
